mbedtls: remove static keyword from certain function pointers

SebastianBoe opened the pull request:
TF-M usually downloads mbedtls and then applies a handful of
patches. These patches were forgotten when Zephyr started to use it's
own mbedtls fork with TF-M.

One of these patches has now been re-applied in this commit. It is
needed for an important code sharing optimization documented in a link
below.

The static keyword is removed from three function pointer
declarations. Aside from allowing the optimization described below
this could theoretically backfire.

There could be other non-static symbols that now trigger aliasing
issues.

The compiler could have trouble optimizing now that it doesn't know of
all symbol usages.

But I believe that in practice this will be OK.

I don't know if there are any plans to upstream this patch to mbedtls
itself.

https://tf-m-user-guide.trustedfirmware.org/docs/technical_references/design_docs/code_sharing.html

Original patch:
https://git.trustedfirmware.org/TF-M/trusted-firmware-m.git/commit/lib/ext/mbedcrypto?id=4a5cc9776ed3b053bac57326931f936fbbc660e9

tamasban committed:
mbedtls: remove static keyword from certain function pointers

TF-M usually downloads mbedtls and then applies a handful of
patches. These patches were forgotten when Zephyr started to use it's
own mbedtls fork with TF-M.

One of these patches has now been re-applied in this commit. It is
needed for an important code sharing optimization documented in a link
below.

The static keyword is removed from three function pointer
declarations. Aside from allowing the optimization described below
this could theoretically backfire.

There could be other non-static symbols that now trigger aliasing
issues.

The compiler could have trouble optimizing now that it doesn't know of
all symbol usages.

But I believe that in practice this will be OK.

I don't know if there are any plans to upstream this patch to mbedtls
itself.

https://tf-m-user-guide.trustedfirmware.org/docs/technical_references/design_docs/code_sharing.html

Original patch:
https://git.trustedfirmware.org/TF-M/trusted-firmware-m.git/commit/lib/ext/mbedcrypto?id=4a5cc9776ed3b053bac57326931f936fbbc660e9

Signed-off-by: Tamas Ban <tamas.ban@arm.com>
Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe commented:
@d3zd3z : This PR sends the patch to the zephyr branch, the previous PR incorrectly applied changes to the master branch.

https://github.com/zephyrproject-rtos/mbedtls/pull/28#issuecomment-957974273

d3zd3z reviewed (changes requested):
Would it be possible to use the commit text, and authorship of the patch we want to apply? You can add [text] tags by the SoB to indicate where you got it from.

SebastianBoe commented:
SoB?

d3zd3z commented:
Signed-off-by

SebastianBoe commented:
It's possible, but there is next-to-none context there, compared to the detailed information in the current git commit message.

So I would prefer to keep the git commit message as-is.

The original git commit message is:

```
+Subject: [PATCH] Enable crypto code sharing between independent binaries
+
+Signed-off-by: Tamas Ban <tamas.ban@arm.com>

```

d3zd3z commented:
Although this isn't a significant amount of code, you really should include the Signed-off-by line from Tamas, and I'd credit him as the author of the original patch. Again, it really isn't as important, since it is just commenting out three lines.

SebastianBoe commented:
Added SoB and changed the author.

Not sure I agree about changing the author, as the original patch is linked to from commit message, but let's just finish this and move on.

SebastianBoe commented:
@d3zd3z : Review has been addressed
